--- a/make-linux.mk
+++ b/make-linux.mk
@@ -268,6 +268,7 @@ ifeq ($(ZT_IA32),1)
 	ZT_USE_X64_ASM_ED25519=0
 endif
 
+ZT_SSO_SUPPORTED=0
 ifeq ($(ZT_SSO_SUPPORTED), 1)
 	ifeq ($(ZT_DEBUG),1)
 		LDLIBS+=zeroidc/target/debug/libzeroidc.a -ldl -lssl -lcrypto
--- a/node/Constants.hpp
+++ b/node/Constants.hpp
@@ -675,4 +675,10 @@
 #define ZT_EXCEPTION_INVALID_SERIALIZED_DATA_INVALID_CRYPTOGRAPHIC_TOKEN 202
 #define ZT_EXCEPTION_INVALID_SERIALIZED_DATA_BAD_ENCODING 203
 
+// Disable Rust build on OpenWrt
+#ifdef OIDC_SUPPORTED
+#undef OIDC_SUPPORTED
+#define OIDC_SUPPORTED 0
+#endif
+
 #endif
